dolphin/task_manager/manager.py

```diff
@@ -67,19 +70,38 @@ def say_hello(self, context, request_spec=None,
             pass
 
     def sync_storage_resource(self, context, storage_id, resource_task):
-        LOG.debug("Received the sync_storage task: {0} request for storage"
-                  " id:{1}".format(resource_task, storage_id))
-        cls = importutils.import_class(resource_task)
-        device_obj = cls(context, storage_id)
-        device_obj.sync()
+        lock = coordination.Lock(storage_id + resource_task)
+        if lock.acquire(False):
+            LOG.debug("Received the sync_storage task: {0} request for storage"
+                      " id:{1}".format(resource_task, storage_id))
+            cls = importutils.import_class(resource_task)
+            device_obj = cls(context, storage_id)
+            device_obj.sync()
+            lock.release()
+        else:
+            LOG.info("%s is rejected for %s because "
+                     "task is already running" % (resource_task, storage_id))
 
     def remove_storage_resource(self, context, storage_id, resource_task):
-        cls = importutils.import_class(resource_task)
-        device_obj = cls(context, storage_id)
-        device_obj.remove()
+        lock = coordination.Lock(storage_id + resource_task)
+        if lock.acquire(False):
+            cls = importutils.import_class(resource_task)
+            device_obj = cls(context, storage_id)
+            device_obj.remove()
+            lock.release()
+        else:
+            LOG.info("%s is rejected for %s because "
+                     "task is already running" % (resource_task, storage_id))
 
     def remove_storage_in_cache(self, context, storage_id):
-        LOG.info('Remove storage device in memory for storage id:{0}'
-                 .format(storage_id))
-        drivers = driver_manager.DriverManager()
-        drivers.remove_driver(context, storage_id)
+        # Use thread lock instead of distributed lock here
+        # Because the data is only available for current thread
+        if thread_lock.acquire(blocking=False):
```

Comment:
What about two threads are removing two different resources concurrently? Say threadA get the lock and delete resourceA, and threadB get lock failed, and then give up, so resourceB cannot be deleted successfully...

Reply:
This could be a problem. Maybe for data in cache, we do not need lock.

Reply:
same doubt.

Reply:
Thread lock has been removed